# Two-Qubit Hadamard Statevector and Normalization

---

## 1. Problem Statement
Construct a 2-qubit quantum circuit initialized in $|00\rangle$, apply Hadamard gates to both qubits, obtain the resulting statevector, display the amplitudes of the computational basis states, and verify that the sum of the corresponding probabilities is equal to 1.

---

## 2. Concept Explanation

### Initial Two-Qubit Computational Basis State
In multi-qubit quantum systems, the state space is formed by the tensor product of individual single-qubit Hilbert spaces. For a 2-qubit system, the computational basis is spanned by $2^2 = 4$ orthonormal basis states:
$$|00\rangle = \begin{pmatrix} 1 \\ 0 \\ 0 \\ 0 \end{pmatrix}, \quad |01\rangle = \begin{pmatrix} 0 \\ 1 \\ 0 \\ 0 \end{pmatrix}, \quad |10\rangle = \begin{pmatrix} 0 \\ 0 \\ 1 \\ 0 \end{pmatrix}, \quad |11\rangle = \begin{pmatrix} 0 \\ 0 \\ 0 \\ 1 \end{pmatrix}$$

By convention in Qiskit, each qubit is initialized into the computational ground state $|0\rangle$, so the initial 2-qubit composite state is:
$$|\psi_{\text{initial}}\rangle = |0\rangle \otimes |0\rangle = |00\rangle$$

### Equal Superposition via Hadamard Gates
The single-qubit Hadamard gate ($H$) transforms the ground state into a symmetric superposition:
$$H = \frac{1}{\sqrt{2}} \begin{pmatrix} 1 & 1 \\ 1 & -1 \end{pmatrix}, \qquad H|0\rangle = \frac{|0\rangle + |1\rangle}{\sqrt{2}}$$

Applying Hadamard gates independently to both qubits represents the tensor product operator $H^{\otimes 2} = H \otimes H$:
$$H^{\otimes 2} = \frac{1}{2} \begin{pmatrix} 1 & 1 & 1 & 1 \\ 1 & -1 & 1 & -1 \\ 1 & 1 & -1 & -1 \\ 1 & -1 & -1 & 1 \end{pmatrix}$$

When applied to the composite ground state $|00\rangle$, it produces an equal superposition across all four computational basis states:
$$|\psi\rangle = (H|0\rangle) \otimes (H|0\rangle) = \left(\frac{|0\rangle + |1\rangle}{\sqrt{2}}\right) \otimes \left(\frac{|0\rangle + |1\rangle}{\sqrt{2}}\right) = \frac{1}{2} \big(|00\rangle + |01\rangle + |10\rangle + |11\rangle\big)$$

### Statevector and Probability Amplitudes
A **statevector** is a complex mathematical vector containing the probability amplitudes associated with each computational basis state of a closed quantum system:
$$|\psi\rangle = \sum_{i=0}^{3} \alpha_i |i\rangle = \alpha_{00}|00\rangle + \alpha_{01}|01\rangle + \alpha_{10}|10\rangle + \alpha_{11}|11\rangle$$

In this equal superposition state:
- Each computational basis state has an identical amplitude:
  $$\alpha_{00} = \alpha_{01} = \alpha_{10} = \alpha_{11} = \frac{1}{2} = 0.5$$
- By the **Born rule**, the probability of observing any particular basis state is the square of the magnitude of its amplitude:
  $$P(|x\rangle) = |\alpha_x|^2 = \left|\frac{1}{2}\right|^2 = \frac{1}{4} = 0.25 \quad (25\%)$$

### Statevector Normalization
The fundamental conservation law of quantum mechanics requires that the sum of the probabilities of all mutually exclusive measurement outcomes must equal unity ($100\%$):
$$\sum_{i} |\alpha_i|^2 = |\alpha_{00}|^2 + |\alpha_{01}|^2 + |\alpha_{10}|^2 + |\alpha_{11}|^2 = \left(\frac{1}{2}\right)^2 + \left(\frac{1}{2}\right)^2 + \left(\frac{1}{2}\right)^2 + \left(\frac{1}{2}\right)^2 = \frac{1}{4} + \frac{1}{4} + \frac{1}{4} + \frac{1}{4} = 1.0$$
where $\alpha_i$ are the complex statevector amplitudes.

---

## 3. Circuit Flow
The sequence of quantum operations is:
$$|00\rangle \xrightarrow{\quad H \text{ on } q_0 \text{ and } H \text{ on } q_1 \quad} |\psi\rangle = \frac{1}{2}\big(|00\rangle + |01\rangle + |10\rangle + |11\rangle\big) \xrightarrow{\quad \text{Statevector Extraction} \quad} \sum |\alpha_i|^2 = 1$$

1. **Initialization**: Both qubits $q_0$ and $q_1$ start in the computational ground state $|00\rangle$.
2. **Hadamard Application**: $H$ is applied to $q_0$ and $q_1$, creating the 2-qubit equal superposition state.
3. **Statevector Extraction**: The 4-dimensional statevector is extracted using `Statevector.from_instruction(qc)`.
4. **Normalization Check**: The squared magnitude of each amplitude is computed and summed to verify $\sum |\alpha_i|^2 = 1$.

---

## 4. Tools and Libraries Used
- **Qiskit**: IBM's open-source quantum computing framework for constructing, transforming, and visualizing quantum circuits (`QuantumCircuit`).
- **qiskit.quantum_info.Statevector**: Class providing exact mathematical representation and direct inspection of quantum statevectors without measurement shots or simulation noise.
- **NumPy**: High-performance scientific computing library used for vectorized amplitude operations, probability calculations, and numerical normalization verification (`np.abs`, `np.sum`, `np.isclose`).

---

## Implementation

### Import Required Libraries
Import `QuantumCircuit` and `Statevector` from Qiskit, along with `numpy` for probability and normalization calculations.

### Construct the 2-Qubit Circuit
Initialize a 2-qubit quantum circuit (`QuantumCircuit(2)`) in the default $|00\rangle$ basis state.

### Apply Hadamard Gates
Apply Hadamard gates to both qubits (`qc.h(0)` and `qc.h(1)`) to prepare the equal superposition state.

### Obtain the Statevector
Extract the 4-element statevector directly from the circuit using `Statevector.from_instruction(qc)`.

### Display Computational-Basis Amplitudes
Iterate over `state.data` using 2-bit binary formatting (`format(i, '02b')`) to display each basis-state amplitude.

### Verify Statevector Normalization
Compute the probability distribution using NumPy (`np.abs(state.data) ** 2`), sum the probabilities (`np.sum(probabilities)`), and verify that $\sum |\alpha_i|^2 = 1.0$ within numerical tolerance.

In [1]:
# Import Required Libraries
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector
import numpy as np

# Construct the 2-Qubit Circuit
qc = QuantumCircuit(2)

# Apply Hadamard Gates
qc.h(0)
qc.h(1)

# Display the Quantum Circuit
print("Quantum Circuit:\n")
print(qc.draw())

# Obtain the Statevector
state = Statevector.from_instruction(qc)

# Display the Statevector
print("\nStatevector:")
print(state)

# Display Computational-Basis Amplitudes
print("\nComputational-Basis Amplitudes:")
for i, amplitude in enumerate(state.data):
    binary_state = format(i, '02b')
    print(f"|{binary_state}⟩ : {amplitude}")

# NumPy Probability Calculation
probabilities = np.abs(state.data) ** 2
print("\nProbabilities:")
for i, prob in enumerate(probabilities):
    binary_state = format(i, '02b')
    print(f"P(|{binary_state}⟩) : {prob:.4f}")

# Verify Statevector Normalization
normalization = np.sum(probabilities)
print(f"\nNormalization (Sum of Probabilities) = {normalization:.4f}")

if np.isclose(normalization, 1.0):
    print("Verification: PASS (Statevector is normalized)")
else:
    print("Verification: FAIL")

/var/folders/yp/078pyxw11mq6g7sd_2ty18z80000gn/T/ipykernel_31203/3476162913.py:1: DeprecationWarning: Using Qiskit with Python 3.9 is deprecated as of the 2.1.0 release. Support for running Qiskit with Python 3.9 will be removed in the 2.3.0 release, which coincides with when Python 3.9 goes end of life.
  from qiskit import QuantumCircuit


Quantum Circuit:

     ┌───┐
q_0: ┤ H ├
     ├───┤
q_1: ┤ H ├
     └───┘

Statevector:
Statevector([0.5+0.j, 0.5+0.j, 0.5+0.j, 0.5+0.j],
            dims=(2, 2))

Computational-Basis Amplitudes:
|00⟩ : (0.4999999999999999+0j)
|01⟩ : (0.4999999999999999+0j)
|10⟩ : (0.4999999999999999+0j)
|11⟩ : (0.4999999999999999+0j)

Probabilities:
P(|00⟩) : 0.2500
P(|01⟩) : 0.2500
P(|10⟩) : 0.2500
P(|11⟩) : 0.2500

Normalization (Sum of Probabilities) = 1.0000
Verification: PASS (Statevector is normalized)


---

## Conclusion

This experiment demonstrates the preparation and mathematical verification of a **2-qubit Hadamard statevector**:

1. **Equal Superposition of Four Basis States**:
   - Applying independent Hadamard gates to both qubits transforms the ground state $|00\rangle$ into an equal superposition:
     $$|\psi\rangle = \frac{1}{2}\big(|00\rangle + |01\rangle + |10\rangle + |11\rangle\big)$$
   - Each computational basis state ($|00\rangle, |01\rangle, |10\rangle, |11\rangle$) has an identical probability amplitude of $0.5 + 0j$.

2. **Equiprobable Measurement Distribution**:
   - The theoretical measurement probability for each of the four basis states is:
     $$P(|00\rangle) = P(|01\rangle) = P(|10\rangle) = P(|11\rangle) = |0.5|^2 = 0.25 \quad (25\%)$$

3. **Normalization Confirmation**:
   - The sum of the squared amplitudes calculated using NumPy is:
     $$\sum_{i} |\alpha_i|^2 = 4 \times (0.2500) = 1.0000$$
   - The verification check passes with high precision (`np.isclose`), conclusively confirming that the statevector represents a valid, normalized pure quantum state.